src/model_lifecycle.cc

```diff
@@ -644,17 +640,23 @@ ModelLifeCycle::UpdateModelConfig(
   model_info->state_reason_.clear();
 
   // Downcast 'Model' to 'TritonModel'.
-  TritonModel* model = (TritonModel*)model_info->model_.get();
+  TritonModel* model = dynamic_cast<TritonModel*>(model_info->model_.get());
   if (model == nullptr) {
     model_info->state_reason_ =
         "Unable to downcast '" + model_id.str() +
         "' from 'Model' to 'TritonModel' during model update.";
     return;
   }
 
-  // Update model instance group.
-  Status status =
-      model->UpdateInstanceGroup(new_model_config, &model_info_lock);
+  // Update model instance group. The 'model_info->mtx_' may be released while
+  // the model is updating its instance group, because no model info from this
+  // model lifecycle object is being accessed by the model during the update,
+  // and model repository manager will ensure no concurrent update on the same
+  // model may take place. Therefore, releasing the lock allows the model to be
+  // accessed which enables inference during update.
+  model_info_lock.unlock();
+  Status status = model->UpdateInstanceGroup(new_model_config);
+  model_info_lock.lock();
```

Comment:
I notice we're now unblocking the lock for a longer section compared to before. Before we passed the lock and only unlocked it for for the call to `SetInstances()`. Can you summarize the difference if any? Or just noticed an improvement to safely unlock for longer?

Reply:
The lock is not necessary, because it cannot stop inference at the first place, and no concurrent model instance update can happen as it is safeguarded by model repository manager. Therefore, the only two places where locking is needed are on the rate limiter and scheduler. They both have their internal lock to safeguard any concurrent access, so the lock is not necessary.